bee-crypto/src/ternary/bigint/u384/mod.rs

```diff
@@ -0,0 +1,482 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+//! This module contains unsigned integers encoded by 384 bits.
+
+mod constants;
+
+pub use constants::{
+    BE_U32_0, BE_U32_1, BE_U32_2, BE_U32_HALF_MAX, BE_U32_HALF_MAX_T242, BE_U32_MAX, BE_U8_0, BE_U8_1, BE_U8_2,
+    BE_U8_MAX, LE_U32_0, LE_U32_1, LE_U32_2, LE_U32_HALF_MAX, LE_U32_HALF_MAX_T242, LE_U32_MAX, LE_U32_MAX_T242,
+    LE_U32_NEG_HALF_MAX_T242, LE_U32_ONLY_T243_OCCUPIED, LE_U8_0, LE_U8_1, LE_U8_2, LE_U8_MAX,
+};
+
+use crate::ternary::bigint::{
+    binary_representation::{BinaryRepresentation, U32Repr, U8Repr},
+    endianness::{BigEndian, LittleEndian},
+    error::Error,
+    overflowing_add::OverflowingAdd,
+    split_integer::SplitInteger,
+    t243, I384, T242, T243,
+};
+
+use bee_ternary::Utrit;
+
+use byteorder::{self, ByteOrder};
+
+use std::{
+    cmp::Ordering,
+    convert::TryFrom,
+    fmt,
+    marker::PhantomData,
+    ops::{Deref, DerefMut},
+};
+
+/// A big integer encoding an unsigned integer with 384 bits.
+///
+/// `T` is usually taken as a `[u32; 12]` or `[u8; 48]`.
+///
+/// `E` refers to the endianness of the digits in `T`. This means that in the case of `[u32; 12]`, if `E == BigEndian`,
+/// that the u32 at position i=0 is considered the most significant digit. The endianness `E` here makes no statement
+/// about the endianness of each single digit within itself (this then is dependent on the endianness of the platform
+/// this code is run on).
+///
+/// For `E == LittleEndian` the digit at the last position is considered to be the most significant.
+#[derive(Clone, Copy)]
+pub struct U384<E, T> {
+    pub(crate) inner: T,
+    _phantom: PhantomData<E>,
+}
+
+impl<E, T> Deref for U384<E, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<E, T> DerefMut for U384<E, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl<E: fmt::Debug, R: BinaryRepresentation, D> fmt::Debug for U384<E, R>
```

Comment:
Are the bounds on the `impl` really necessary? Maybe I'm wrong, but I don't think so.

Reply:
Removed